File: ios/FAHYBRIKCore/Workout/BeltWorkClock.swift
import Foundation

/// Reloj de TRABAJO en cinta FTMS. El de la sesión (lap / EMOM / AMRAP) es
/// de pared. Éste solo suma cuando la máquina manda velocidad.
enum BeltWorkClock {
    static let minMovingKmh: Double = 0.5

    enum Surface {
        case ftms
        case other
    }

    enum Window {
        case work
        case recovery
        case countIn
        case format
    }

    static func isMoving(_ speedKmh: Double?) -> Bool {
        (speedKmh ?? 0) > minMovingKmh
    }

    /// `beltMoving == nil`: no hay feed FTMS. El reloj de trabajo no se aplica.
    static func applies(surface: Surface, window: Window, beltMoving: Bool?) -> Bool {
        surface == .ftms && window == .work && beltMoving != nil
    }

    static func workTick(wallDt: TimeInterval, surface: Surface,
                         window: Window, beltMoving: Bool?) -> TimeInterval {
        guard applies(surface: surface, window: window, beltMoving: beltMoving) else {
            return wallDt
        }
        return (beltMoving ?? false) ? wallDt : 0
    }
}

File: shared/domain/belt-work-clock.ts
// Reloj de TRABAJO en cinta FTMS. El de la sesión (lap / EMOM / AMRAP) es
// otro: es de pared. Éste solo suma cuando la máquina manda velocidad.

export const MIN_BELT_MOVING_KMH = 0.5;

export type BeltClockSurface = 'ftms' | 'other';
export type BeltClockWindow = 'work' | 'recovery' | 'count_in' | 'format';

export function beltIsMoving(speedKmh: number | null | undefined): boolean {
  return (speedKmh ?? 0) > MIN_BELT_MOVING_KMH;
}

/**
 * `beltMoving === null`: no hay feed FTMS (cinta tonta, reloj, calle).
 * El reloj de trabajo no se aplica.
 */
export function beltWorkApplies(args: {
  surface: BeltClockSurface;
  window: BeltClockWindow;
  beltMoving: boolean | null;
}): boolean {
  return args.surface === 'ftms' && args.window === 'work' && args.beltMoving !== null;
}

/** dt que cuenta como trabajo. 0 si la banda no manda velocidad. */
export function beltWorkTick(args: {
  wallDt: number;
  surface: BeltClockSurface;
  window: BeltClockWindow;
  beltMoving: boolean | null;
}): number {
  if (!beltWorkApplies(args)) return args.wallDt;
  return args.beltMoving ? args.wallDt : 0;
}

/** Ritmo (s/km) desde km/h. Parado no tiene ritmo. */
export function paceSecPerKmFromKmh(speedKmh: number | null | undefined): number | null {
  if (!beltIsMoving(speedKmh) || speedKmh == null) return null;
  return Math.round(3600 / speedKmh);
}
